### PR TITLE
feat: add custom driver for SQL Server

### PR DESCRIPTION
**Pull request by enocom**

### CODE REVIEW

**Review by kurtisvg — CHANGES_REQUESTED** [review]

File: sqlserver/sqlserver.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package sqlserver provides a Cloud SQL SQL Server driver that works with the
+// database/sql package.
+package sqlserver
```
[the remark this answers is not in the document]
Comment (enocom) [reply sqlserver/sqlserver.go:17]: Done with `mssql`.

File: sqlserver/sqlserver.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+// "sqlserver://user:password@localhost?database=mydb&cloudsql=my-proj:us-central1:my-inst"
+//
+// For details, see
+// https://github.com/denisenkom/go-mssqldb#the-connection-string-can-be-specified-in-one-of-three-formats
+func (s *sqlserverDriver) Open(name string) (driver.Conn, error) {
+	_, res, err := msdsn.Parse(name)
+	if err != nil {
+		return nil, err
+	}
+	d, err := cloudsqlconn.NewDialer(context.Background(), s.opts...)
```
[the remark this answers is not in the document]
Comment (enocom) [reply sqlserver/sqlserver.go:74]: We can do this, but will have to return an error in the `RegisterDriver` call.

File: README.md [hunk 3]
```diff
@@ -77,6 +77,10 @@ if connErr != nil {
 defer conn.Close(ctx)
 ```
 
+#### SQL Server Support
+
+See [below](#SQL-Server).
```
[the remark this answers is not in the document]
Comment (enocom) [reply README.md]: Updated the wording. PTAL.

File: README.md [hunk 4]
```diff
@@ -157,6 +161,37 @@ func Connect() {
...
+    "cloud.google.com/go/cloudsqlconn"
+    "cloud.google.com/go/cloudsqlconn/sqlserver"
+)
+
+func Connect() {
+    // Without any options:
+    sqlserver.RegisterDriver("cloudsql-sqlserver")
+
+    // Or, with options:
+    // sqlserver.RegisterDriver("cloudsql-sqlserver", cloudsqlconn.WithCredentialsFile("key.json"))
```
[the remark this answers is not in the document]
Comment (enocom) [reply README.md]: Fixed.

File: sqlserver/mssql/mssql.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+func (s *sqlserverDriver) Open(name string) (driver.Conn, error) {
+	_, res, err := msdsn.Parse(name)
+	if err != nil {
+		return nil, err
+	}
+	c, err := mssqldb.NewConnector(name)
+	if err != nil {
+		return nil, err
+	}
+	connName := res["cloudsql"]
```
Comment (kurtisvg) [inline sqlserver/mssql/mssql.go:88]: Is this the arg for the Cloud SQL instance name? Can it be passed in as the hostname instead? e.g. `sqlserver://user:password@PROJECT:REGION:INST?database=mydb`

**Review by kurtisvg — CHANGES_REQUESTED** [review]

File: sqlserver/mssql/mssql.go (hunk 5 above)
Comment (enocom) [reply sqlserver/mssql/mssql.go:88]: Yes. It we pass it as hostname, the DSN parsing library fails to parse it. It sees the final colon as indicating port.

Comment (enocom) [reply sqlserver/mssql/mssql.go:88]: With the current approach, we support all three DSN formats, albeit with this `cloudsql` parameter. Formats are documented [here](https://github.com/denisenkom/go-mssqldb#the-connection-string-can-be-specified-in-one-of-three-formats).

**Review by kurtisvg — APPROVED** [review]

**Pull request merged**